pkg/kubelet/kubelet.go

```diff
@@ -557,11 +558,78 @@ func (kl *Kubelet) Run(updates <-chan PodUpdate) {
 	kl.syncLoop(updates, kl)
 }
 
+func (kl *Kubelet) initialNodeStatus() (*api.Node, error) {
+	node := &api.Node{
+		ObjectMeta: api.ObjectMeta{Name: kl.hostname},
+		Status: api.NodeStatus{
+			Capacity: api.ResourceList{
+				api.ResourceCPU:    *resource.NewMilliQuantity(1000, resource.DecimalSI),
+				api.ResourceMemory: resource.MustParse("3Gi"),
+			},
+		},
+	}
+	if kl.cloud != nil {
+		instances, ok := kl.cloud.Instances()
+		if !ok {
+			return nil, fmt.Errorf("failed to get instances from cloud provider")
+		}
+		// TODO(roberthbailey): Inspect the local machine rather than calling out
+		// to the cloud provider.
+		resources, err := instances.GetNodeResources(kl.hostname)
+		if err != nil {
+			return nil, fmt.Errorf("failed to get node resources from cloud provider: %v", err)
+		}
+		if resources != nil {
+			node.Status.Capacity = resources.Capacity
+		}
+		instanceID, err := instances.ExternalID(kl.hostname)
+		if err != nil {
+			return nil, fmt.Errorf("failed to get instance ID from cloud provider: %v", err)
+		}
+		node.Spec.ExternalID = instanceID
+		nodeAddresses, err := instances.NodeAddresses(kl.hostname)
```

Comment:
I thought I copied this directly from populateAddresses but it sounds like I missed a bit. I'll go back and check. 

Addresses are part of the node status. So it should be possible for the node to notice that it's address changed and send that as part of the regular status update to the master (It doesn't do that yet).